app/components/user/TwoFactor/configure.js

```diff
@@ -0,0 +1,325 @@
+      return (
+        <Panel>
+          <Panel.Section>
+            {this.props.viewer.totp && (
+              <Panel className="orange border-orange">
+                <Panel.Section>
+                  <strong>Youʼre about to reconfigure two-factor authentication.</strong> This will replace your existing two-factor authentication applications and recovery codes.
+                </Panel.Section>
+              </Panel>
+            )}
+
+            <p>To {this.props.viewer.totp ? 'reconfigure' : 'activate'} two-factor authentication, scan this QR Code with your authenticator application.</p>
+
+            <QRCode
+              renderAs="svg"
+              fgColor="currentColor"
+              bgColor="transparent"
+              width="300"
+              height="auto"
+              className="block my4 mx-auto"
+              style={{
+                maxWidth: '100%'
+              }}
+              value={this.state.provisioningUri}
+            />
+
+            <p>If you need an authenticator application, some good options include {AUTHENTICATOR_LIST}.</p>
+          </Panel.Section>
+
+          <Panel.Section>
+            <form onSubmit={this.handleActivateSubmit}>
+              <FormTextField
+                label="Two-factor authentication code"
+                autoFocus={true}
+                placeholder="123456"
+                onChange={this.handleCodeChange}
+              />
+
+              <Button
+                className="col-12"
+                theme="success"
+                loading={this.state.activatingTOTP && 'Activating…'}
+              >
+                Activate
+              </Button>
+            </form>
+          </Panel.Section>
+        </Panel>
+      );
+    } else if (this.state.activatedTOTP) {
+      return (
+        <Panel>
+          <Panel.Section>
+            <p>Congratulations! Two-factor authentication is now configured for your account.</p>
+
+            <Button
+              theme="success"
+              link="/user/two-factor"
+            >
+              Back to Two-Factor Authentication Settings
+            </Button>
+          </Panel.Section>
+        </Panel>
+      );
+    }
+  }
+
+  componentDidMount() {
+    // Immediately go generate a token
+    commitMutation(this.props.relay.environment, {
+      mutation: graphql`
+        mutation configureCreateMutation($input: TOTPCreateInput!) {
+          totpCreate(input: $input) {
+            clientMutationId
+            provisioningUri
+            totp {
+              id
+            }
+          }
+        }
+      `,
+      variables: { input: {} },
+      onCompleted: this.handleCreateMutationComplete,
+      onError: this.handleCreateMutationError
+    });
+  }
+
+  componentWillUnmount() {
+    // When unmounting, if we haven't finished activating, try to delete the TOTP record
+    if (this.state.totpId && this.state.generatedTOTP && !this.state.activatedTOTP) {
```

Comment:
Will this block an unmount? What happens if this query likes to take it's time?

Reply:
Good question! This won’t block an unmount or navigation; it’s entirely async. It doesn’t really matter much if this succeeds or not (a non-activated TOTP configuration simply has no effect), this is more or less a best effort to clean up after ourselves.

This can be enhanced further (perhaps even obviated) by having a cleanup worker of some sort to delete them automatically (which we’ve discussed before, but I don’t consider high priority to implement until the rest of the shape of this is better understood!).